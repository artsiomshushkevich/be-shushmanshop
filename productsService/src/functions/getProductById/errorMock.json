{
    "httpMethod": "GET",
    "headers": {
        "Content-Type": "application/json"
    },
    "pathParameters": {
        "id": "9999"
    },
    "body": null
}
