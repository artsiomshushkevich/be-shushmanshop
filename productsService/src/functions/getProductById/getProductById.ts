import { v4 as uuidv4 } from 'uuid';
import type { ValidatedEventAPIGatewayProxyEvent } from '@libs/api-gateway';
import { formatJSONResponse } from '@libs/api-gateway';
import type { HttpErrorBody } from '@localtypes/httpErrorBody';
import { HttpStatusCodes } from '@localtypes/httpStatusCodes';
import { productsService } from '@services/products';
import { createHttpErrorResponseObject } from '@utils/createHttpErrorResponseObject/createHttpErrorResponseObject';

export const getProductById: ValidatedEventAPIGatewayProxyEvent<null> = async (event) => {
    try {
        const { id } = event.pathParameters;

        console.log(
            'Getting specific product started. The product id is %s. Parameters are %s',
            id,
            JSON.stringify(event.queryStringParameters)
        );

        const product = await productsService.getById(id);

        if (product) {
            return formatJSONResponse(product);
        } else {
            const notFoundMesssage = 'Product with such id not found!';
            const notFoundUuid = uuidv4();

            console.log(`${notFoundMesssage} The product id is ${id}. UUID: ${notFoundUuid}.`);

            const errorBody: HttpErrorBody = {
                message: notFoundMesssage,
                uuid: notFoundUuid
            };

            return formatJSONResponse(errorBody, HttpStatusCodes.NotFound);
        }
    } catch (e) {
        const errorUuid = uuidv4();

        console.error('Getting specific product failed. UUID: %s', errorUuid, e);

        return formatJSONResponse(
            createHttpErrorResponseObject('Smth went wrong!', errorUuid),
            HttpStatusCodes.ServerError
        );
    }
};
