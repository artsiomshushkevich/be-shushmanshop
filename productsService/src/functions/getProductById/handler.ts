import { v4 as uuidv4 } from 'uuid';
import type { ValidatedEventAPIGatewayProxyEvent } from '@libs/api-gateway';
import { formatJSONResponse } from '@libs/api-gateway';
import { HttpErrorBody } from '@localtypes/httpErrorBody';
import { HttpStatusCodes } from '@localtypes/httpStatusCodes';
import { productsModel } from '@models/products';

export const main: ValidatedEventAPIGatewayProxyEvent<null> = async (event) => {
  const { id } = event.pathParameters;

  const product = await productsModel.getById(id);

  if (product) {
    return formatJSONResponse(product);
  } else {
    const errorBody: HttpErrorBody = {
      message: 'Product with such id not found!',
      uuid: uuidv4()
    };

    return formatJSONResponse(errorBody, HttpStatusCodes.NotFound);
  }
};