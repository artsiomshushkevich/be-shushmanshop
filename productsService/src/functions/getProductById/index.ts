import { handlerPath } from '@libs/handler-resolver';
import type { LamdaConfig } from '@libs/types';

export const getProductById: LamdaConfig = {
  handler: `${handlerPath(__dirname)}/handler.main`,
  events: [
    {
      http: {
        method: 'get',
        path: 'product/{id}'
      }
    }
  ]
};
