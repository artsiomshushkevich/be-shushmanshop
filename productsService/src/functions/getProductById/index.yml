getProductById:
    handler: src/functions/getProductById/handler.main
    events:
        - http:
              method: GET
              path: /product/{id}
