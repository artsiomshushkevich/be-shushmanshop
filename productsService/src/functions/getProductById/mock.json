{
    "httpMethod": "GET",
    "headers": {
        "Content-Type": "application/json"
    },
    "pathParameters": {
        "id": "2"
    },
    "body": null
}
