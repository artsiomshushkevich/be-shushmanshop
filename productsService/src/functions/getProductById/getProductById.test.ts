import { getProductById } from './getProductById';
import event from './mock.json';
import errorEvent from './errorMock.json';

jest.mock('uuid', () => ({ v4: () => '916eb106-2ecc-4ac9-af2c-044f75220c5b' }));

describe('getProductById', () => {
    // TODO: stab DB utils later
    test.skip('Should return correct response', async () => {
        // @ts-ignore
        const response = await getProductById(event);

        expect(response).toMatchSnapshot();
    });

    test.skip('Should return 404 error response', async () => {
        // @ts-ignore
        const response = await getProductById(errorEvent);

        expect(response).toMatchSnapshot();
    });
});
