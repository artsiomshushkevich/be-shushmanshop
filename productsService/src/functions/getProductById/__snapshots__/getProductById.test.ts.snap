// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getProductById Should return 404 error response 1`] = `
{
  "body": "{"message":"Product with such id not found!","uuid":"916eb106-2ecc-4ac9-af2c-044f75220c5b"}",
  "statusCode": 404,
}
`;

exports[`getProductById Should return correct response 1`] = `
{
  "body": "{"id":"2","title":"Volkswagen","description":"Golf E","price":25000}",
  "statusCode": 200,
}
`;
